lib/ws-channel.js

```diff
@@ -0,0 +1,250 @@
+  this._callbacksPending++;
+
+  // Reference the socket (even if was referenced) if we're waiting for a
+  // response.
+  if (this._callbacksPending == 1 && this._socket && this._unref)
+    this._socket.ref();
+
+  var packet = {
+    type: REQ,
+    seqno: seqno,
+    data: message
+  };
+
+  this._send(packet);
+};
+
+
+WebsocketChannel.prototype.address = function address() {
+  return this.uri && url.parse(this.uri);
+};
+
+
+WebsocketChannel.prototype.getToken = function getToken() {
+  return this._token;
+};
+
+
+WebsocketChannel.prototype.unref = function unref() {
+  this._unref = true;
+  if (this._socket && this._callbacksPending === 0)
+    this._socket.unref();
+};
+
+
+WebsocketChannel.prototype._onOpen = function() {
+  this._socket = this._websocket._socket;
+
+  if (this._unref && this._callbacksPending === 0)
+    this._socket.unref();
+
+  this._flushSendQueue();
+};
+
+
+WebsocketChannel.prototype._send = function(message) {
+  this._sendQueue.push(JSON.stringify(message));
+  if (this._isOpen())
+    this._flushSendQueue();
+};
+
+
+WebsocketChannel.prototype._handleMessage = function(s) {
+  var self = this;
+
+  try {
+    var packet = JSON.parse(s);
+  } catch (err) {
+    return this._handleError(err);
+  }
+
+  switch (packet.type) {
+    case NOT:
+      return this._onNotification(packet.data);
+
+    case REQ:
+      return this._onRequest(packet.data, sendReply);
+
+    case RSP:
+      var callback = this._callbacks[packet.seqno];
+      if (!callback)
+        return this._handleError(new Error('Unrecognized response sequence number'));
+
+      delete this._callbacks[packet.seqno];
+      this._callbacksPending--;
+
+      // Unref the socket if the channel was unref'ed and we're no longer
+      // awaiting any responses.
+      if (this._unref && this._callbacksPending === 0)
+        this._socket.unref();
+
+      return callback(packet.data);
+
+    default:
+      return this._handleError(new Error('Invalid message received'));
+  }
+
+  function sendReply(message) {
+    packet.type = RSP;
+    packet.data = message;
+    self._send(packet);
+  }
+};
+
+
+WebsocketChannel.prototype._isOpen = function() {
+  return this._websocket && this._websocket.readyState === Websocket.OPEN;
+};
+
+
+WebsocketChannel.prototype._flushSendQueue = function() {
+  var self = this;
+  var sendQueue = this._sendQueue;
+  var s;
+
+  while ((s = this._sendQueue.shift()) !== undefined) {
+    this._websocket.send(s, function(err) {
+      if (err)
+        self._handleError(err);
+    });
+  }
+};
+
+
+WebsocketChannel.prototype._handleError = function(err) {
+  if (this._error)
+    return;
+
+  this._error = err;
+  this._websocket.close();
+
+  this.emit('error', err);
+};
+
+
+WebsocketChannel.create = function create(onRequest, onNotification) {
+  return new WebsocketChannel(onRequest, onNotification);
+};
+
+
+WebsocketChannel.connect = function create(onRequest, onNotification, uri, token) {
```

Comment:
That's the wrong function name, and it pushes over 80 columns. might as well remove the names from the function definitions everywhere, its half and half at the moment (yes, sorry, the existing code was naming functions... its a transition period ATM).